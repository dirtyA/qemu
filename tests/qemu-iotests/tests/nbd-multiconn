#!/usr/bin/env bash
# group: rw auto quick
#
# Test that qemu-nbd MULTI_CONN works
#
# Copyright (C) 2018-2021 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

seq="$(basename $0)"
echo "QA output created by $seq"

status=1 # failure is the default!

_cleanup()
{
    _cleanup_test_img
    nbd_server_stop
}
trap "_cleanup; exit \$status" 0 1 2 3 15

# get standard environment, filters and checks
cd ..
. ./common.rc
. ./common.filter
. ./common.nbd

_supported_fmt qcow2
_supported_proto file
_supported_os Linux
_require_command QEMU_NBD

# Parallel client connections are easier with libnbd than qemu-io:
if ! (nbdsh --version) >/dev/null 2>&1; then
    _notrun "nbdsh utility required, skipped this test"
fi

echo
echo "=== Initial image setup ==="
echo

_make_test_img 4M
$QEMU_IO -c 'w -P 1 0 2M' -c 'w -P 2 2M 2M' "$TEST_IMG" | _filter_qemu_io
IMG="driver=nbd,server.type=unix,server.path=$nbd_unix_socket"
nbd_server_start_unix_socket -f qcow2 -e 5 "$TEST_IMG"

echo
echo "=== Demonstrate parallel clients ==="
echo

export nbd_unix_socket
nbdsh -c '
import os
sock = os.getenv("nbd_unix_socket")
h = []

for i in range(3):
  h.append(nbd.NBD())
  h[i].connect_unix(sock)
  assert h[i].can_multi_conn()

buf1 = h[0].pread(1024 * 1024, 0)
if buf1 != b"\x01" * 1024 * 1024:
  print("Unexpected initial read")
buf2 = b"\x03" * 1024 * 1024
h[1].pwrite(buf2, 0)
h[2].flush()
buf1 = h[0].pread(1024 * 1024, 0)
if buf1 == buf2:
  print("Flush appears to be consistent across connections")

for i in range(3):
  h[i].shutdown()
'

# success, all done
echo '*** done'
rm -f $seq.full
status=0
