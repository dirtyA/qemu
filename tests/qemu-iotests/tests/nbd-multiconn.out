QA output created by nbd-multiconn

=== Initial image setup ===

Formatting 'TEST_DIR/t.IMGFMT', fmt=IMGFMT size=4194304
wrote 2097152/2097152 bytes at offset 0
2 MiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)
wrote 2097152/2097152 bytes at offset 2097152
2 MiB, X ops; XX:XX:XX.X (XXX YYY/sec and XXX ops/sec)

=== Demonstrate parallel clients ===

Flush appears to be consistent across connections
*** done
